### PR TITLE
PG unify/cleanup tf vs torch and PG functionality test cases (tf + torch).

### PR DESCRIPTION
**Pull request by sven1977**
Unifying the code for PGTrainer/Policy wrt tf vs torch.

Unifying the code for PGTrainer/Policy wrt tf vs torch.
Adding loss function test cases for the PGAgent (confirm equivalence of tf and torch).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: rllib/agents/pg/__init__.py [hunk 1]
```diff
@@ -1,6 +1,11 @@
 from ray.rllib.agents.pg.pg import PGTrainer, DEFAULT_CONFIG
+from ray.rllib.agents.pg.tf_pg_policy import tf_pg_loss, \
+    post_process_advantages
+from ray.rllib.agents.pg.torch_pg_policy import torch_pg_loss
 from ray.rllib.utils import renamed_agent
 
+# `Agent`=Legacy: Use PGTrainer instead.
 PGAgent = renamed_agent(PGTrainer)
```
Comment (ericl) [inline rllib/agents/pg/__init__.py]: Btw we can probably start removing these, they've been deprecated for a while now.

File: rllib/agents/pg/torch_pg_policy.py [hunk 2]
```diff
@@ -3,38 +3,37 @@
...
+    """The basic policy gradients loss."""
     logits, _ = model.from_batch(train_batch)
     action_dist = dist_class(logits, model)
     log_probs = action_dist.logp(train_batch[SampleBatch.ACTIONS])
-    # save the error in the policy object
-    policy.pi_err = -train_batch[Postprocessing.ADVANTAGES].dot(
-        log_probs.reshape(-1))
+    # Save the error in the policy object.
+    # policy.pi_err = -train_batch[Postprocessing.ADVANTAGES].dot(
+    # log_probs.reshape(-1)) / len(log_probs)
```
Comment (ericl) [inline rllib/agents/pg/torch_pg_policy.py]: Remove this commented out code?

File: rllib/agents/pg/torch_pg_policy.py [hunk 3]
```diff
@@ -3,38 +3,37 @@
 from __future__ import print_function
 
 import ray
-from ray.rllib.evaluation.postprocessing import compute_advantages, \
-    Postprocessing
+from ray.rllib.agents.pg.tf_pg_policy import post_process_advantages
```
Comment (ericl) [inline rllib/agents/pg/torch_pg_policy.py]: Hm should we be importing tf policy from torch files? I guess it's ok as long as it doesn't fail if tensorflow isn't installed.

File: rllib/tests/agents/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
```
Comment (ericl) [inline rllib/tests/agents/__init__.py]: We can start droppin these now that Py2 support is removed.

File: rllib/tests/agents/functionality/test_pg_functionality.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
+from __future__ import absolute_import
```
Comment (ericl) [inline rllib/tests/agents/functionality/test_pg_functionality.py]: How about we put this file in `agents/pg` to be close to the code? With bazel the tests can go anywhere as long as they have an entry in the BUILD file.

File: rllib/tests/agents/functionality/test_pg_functionality.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+import ray
+import ray.rllib.agents.pg as pg
+from ray.rllib.evaluation.postprocessing import Postprocessing
+from ray.rllib.models.tf.tf_action_dist import Categorical
+from ray.rllib.models.torch.torch_action_dist import TorchCategorical
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils import check, fc
+
+
+class TestPGFunctionality(unittest.TestCase):
```
Comment (ericl) [inline rllib/tests/agents/functionality/test_pg_functionality.py]: TestPG? I don't think `functionality` adds a lot of description here.

File: rllib/tests/test_multi_agent_env.py [hunk 7]
```diff
@@ -8,7 +8,7 @@
 
 import ray
 from ray.rllib.agents.pg import PGTrainer
-from ray.rllib.agents.pg.pg_policy import PGTFPolicy
+from ray.rllib.agents.pg.tf_pg_policy import PGTFPolicy
```
Comment (ericl) [inline rllib/tests/test_multi_agent_env.py]: How about we go with `<alg>_<framework>_policy` as a standard for naming, like `a3c_tf_policy`? That way the policy files always end up near each other in alphabetical sort.

File: rllib/tests/agents/functionality/test_pg_functionality.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        )
+        expected_logits = policy.model._last_output
+        expected_logp = TorchCategorical(expected_logits, policy.model).logp(
+            train_batch[SampleBatch.ACTIONS]
+        )
+        expected_loss = -np.mean(
+            expected_logp.detach().numpy() *
+            train_batch[Postprocessing.ADVANTAGES].numpy()
+        )
+        check(results.detach().numpy(), expected_loss, decimals=4)
```
Comment (ericl) [inline rllib/tests/agents/functionality/test_pg_functionality.py]: I guess other possible reasons torch may still differ from TF include the model initialization, and the optimizer step. Should we test the model starting weights and weights after the step are also identical?

This may be a bit difficult since the models are randomly initialized.

File: rllib/agents/pg/torch_pg_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/pg/torch_pg_policy.py]: Was thinking about making it optional to use either the sum or the mean.

File: rllib/tests/agents/functionality/test_pg_functionality.py (hunk 5 above)
Comment (sven1977) [reply rllib/tests/agents/functionality/test_pg_functionality.py]: As long as we do the location of tests consistently, it's ok with me. Will start cleaning this up.

File: rllib/agents/pg/torch_pg_policy.py (hunk 3 above)
Comment (sven1977) [reply rllib/agents/pg/torch_pg_policy.py]: It won't fail b/c we are using try_import_tf(). And the function itself that we are importing has no tf-dependency in it.
But it's a valid question: Should this function be in the tf-specific file then?

File: rllib/tests/agents/functionality/test_pg_functionality.py (hunk 8 above)
Comment (sven1977) [reply rllib/tests/agents/functionality/test_pg_functionality.py]: Yeah, unless we overwrite the weights with `set_weights()` at the beginning of the test case.

**Review by ericl — APPROVED** [review]

File: rllib/agents/pg/tests/test_pg.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
+import numpy as np
```
Comment (ericl) [inline rllib/agents/pg/tests/test_pg.py]: Needs a BUILD entry

Comment (sven1977) [reply rllib/agents/pg/tests/test_pg.py]: Done, re-testing ...

**Comment by sven1977** [comment]
Thanks!
